File: java/data-structure/src/main/java/org/example/ArrayDeque.java
package org.example;

/**
 * 数组实现双端队列
 *
 * @param <T> 元素类型
 */
public class ArrayDeque<T> implements Deque<T> {
    private static final int CAPACITY = 1 << 3;
    private T[] element;
    private int head;
    private int tail;
    private int size;

    public ArrayDeque() {
        this(CAPACITY);
    }

    public ArrayDeque(int capacity) {
        if (capacity < 0) {
            throw new RuntimeException("必须为非负数");
        }
        element = newArray(multipleOfTwo(capacity));
    }

    private static <T> T[] newArray(int capacity) {
        return (T[]) new Object[capacity];
    }

    private static int multipleOfTwo(int capacity) {
        for (int i = 1; i <= 1 << 4; i <<= 1) {
            capacity |= capacity >>> i;
        }
        return ++capacity < 0 ? capacity >>> 1 : capacity;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size() == 0;
    }

    @Override
    public void clear() {
        int mask = element.length - 1;
        for (int i = head; i != tail; i = i + 1 & mask) {
            element[i] = null;
        }
        head = tail = size = 0;
    }

    @Override
    public boolean add(T t) {
        return offerLast(t);
    }

    @Override
    public boolean contains(T t) {
        int mask = element.length - 1;
        for (int i = head; i != tail; i = i + 1 & mask) {
            if (t == element[i] || t != null && t.equals(element[i])) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean remove(T t) {
        while (!isEmpty()) {
            T poll = pollFirst();
            if (t == poll || t != null && t.equals(poll)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean offer(T t) {
        return offerLast(t);
    }

    @Override
    public boolean push(T t) {
        return offerFirst(t);
    }

    @Override
    public T peek() {
        return peekFirst();
    }

    @Override
    public T pop() {
        return pollFirst();
    }

    @Override
    public T poll() {
        return pollFirst();
    }

    @Override
    public boolean offerFirst(T t) {
        head = head - 1 & element.length - 1;
        element[head] = t;
        size++;
        adjustCapacityCheck(true);
        return true;
    }

    @Override
    public boolean offerLast(T t) {
        element[tail] = t;
        tail = tail + 1 & element.length - 1;
        size++;
        adjustCapacityCheck(true);
        return true;
    }

    @Override
    public T peekFirst() {
        check();
        return element[head];
    }

    @Override
    public T peekLast() {
        check();
        return element[tail - 1 & element.length - 1];
    }

    @Override
    public T pollFirst() {
        check();
        T t = element[head];
        element[head] = null;
        head = head + 1 & element.length - 1;
        size--;
        adjustCapacityCheck(false);
        return t;
    }

    @Override
    public T pollLast() {
        check();
        tail = tail - 1 & element.length - 1;
        T t = element[tail];
        element[tail] = null;
        size--;
        adjustCapacityCheck(false);
        return t;
    }

    @Override
    public Iterator<T> iterator() {
        return new Iter();
    }

    private void check() {
        if (size <= 0) {
            throw new RuntimeException("集合为空");
        }
    }

    private void adjustCapacityCheck(boolean isAdd) {
        int n = element.length;
        if (isAdd && size == n) {
            adjustCapacity(n == 0 ? 1 : n * 2);
        } else if (!isAdd && size == n / 4) {
            adjustCapacity(n == 1 ? 1 : n / 2);
        }
    }

    private void adjustCapacity(int newCapacity) {
        T[] old = element;
        element = newArray(newCapacity);
        int mask = old.length - 1;
        for (int i = 0; i < size; i++) {
            element[i] = old[head + i & mask];
        }
        head = 0;
        tail = size;
    }

    private class Iter implements Iterator<T> {
        private final int fence = tail;
        private final int mask = element.length - 1;
        private int cursor = head;

        @Override
        public boolean hasNext() {
            return cursor != fence;
        }

        @Override
        public T next() {
            if (!hasNext()) {
                throw new RuntimeException("没有下一个元素");
            }
            T t = element[cursor];
            cursor = cursor + 1 & mask;
            return t;
        }
    }
}

File: csharp/data-structure/ArrayDeque.cs
namespace Org.Example
{
    public class ArrayDeque<T> : IDeque<T>, IList<T>
    {
        private static readonly int CAPACITY = 1 << 3;
        private T[] element;
        private int head;
        private int tail;

        public ArrayDeque() : this(CAPACITY) { }

        public ArrayDeque(int capacity)
        {
            if (capacity < 0)
            {
                throw new Exception("必须为非负数");
            }
            element = new T[MultipleOfTwo(capacity)];
        }

        public int Count { get; private set; }

        public bool IsEmpty => Count == 0;

        public void Clear()
        {
            int mask = element.Length - 1;
            for (int i = head; i != tail; i = i + 1 & mask)
            {
                element[i] = default;
            }
            head = tail = Count = 0;
        }

        public bool Add(T item) => OfferLast(item);

        public bool Contains(T item) => IndexOf(item) >= 0;

        public bool Remove(T item)
        {
            int index = IndexOf(item);
            if (index < 0)
            {
                return false;
            }
            RemoveAt(index);
            return true;
        }

        bool IStack<T>.Push(T item) => OfferFirst(item);

        T IStack<T>.Peek() => PeekFirst();

        T IStack<T>.Pop() => PollFirst();

        bool IQueue<T>.Offer(T item) => OfferLast(item);

        T IQueue<T>.Peek() => PeekFirst();

        T IQueue<T>.Poll() => PollFirst();

        public bool OfferFirst(T item)
        {
            head = head - 1 & element.Length - 1;
            element[head] = item;
            Count++;
            AdjustCapacityCheck(true);
            return true;
        }

        public bool OfferLast(T item)
        {
            element[tail] = item;
            tail = tail + 1 & element.Length - 1;
            Count++;
            AdjustCapacityCheck(true);
            return true;
        }

        public T PeekFirst()
        {
            Check();
            return element[head];
        }

        public T PeekLast()
        {
            Check();
            return element[tail - 1 & element.Length - 1];
        }

        public T PollFirst()
        {
            Check();
            T item = element[head];
            element[head] = default;
            head = head + 1 & element.Length - 1;
            Count--;
            AdjustCapacityCheck(false);
            return item;
        }

        public T PollLast()
        {
            Check();
            tail = tail - 1 & element.Length - 1;
            T item = element[tail];
            element[tail] = default;
            Count--;
            AdjustCapacityCheck(false);
            return item;
        }

        public T this[int index]
        {
            get
            {
                index = IndexOffset(index);
                Check(index, false);
                return element[index];
            }
            set
            {
                index = IndexOffset(index);
                Check(index, false);
                element[index] = value;
            }
        }

        public bool Insert(int index, T item)
        {
            index = IndexOffset(index);
            Check(index, true);
            int mask = element.Length - 1;
            for (int i = tail; i != index; i = i - 1 & mask)
            {
                element[i] = element[i - 1 & mask];
            }
            tail = tail + 1 & mask;
            element[index] = item;
            Count++;
            AdjustCapacityCheck(true);
            return true;
        }

        public T RemoveAt(int index)
        {
            index = IndexOffset(index);
            Check(index, false);
            T item = element[index];
            int mask = element.Length - 1;
            for (int i = index; i != tail; i = i + 1 & mask)
            {
                element[i] = element[i + 1 & mask];
            }
            tail = tail - 1 & mask;
            element[tail] = default;
            Count--;
            AdjustCapacityCheck(false);
            return item;
        }

        public int IndexOf(T item)
        {
            int index = 0;
            int mask = element.Length - 1;
            for (int i = head; i != tail; i = i + 1 & mask, index++)
            {
                if (Equals(item, element[i]))
                {
                    return index;
                }
            }
            return -1;
        }

        public IEnumerator<T> GetEnumerator() => new Etor(this);

        private static int MultipleOfTwo(int capacity)
        {
            for (int i = 1; i <= 1 << 4; i <<= 1)
            {
                capacity |= (int)((uint)capacity >> i);
            }
            return ++capacity < 0 ? (int)((uint)capacity >> 1) : capacity;
        }

        private int IndexOffset(int index)
        {
            if (index < 0)
            {
                throw new Exception("索引必须为非负数");
            }
            return index + head & element.Length - 1;
        }

        private void Check()
        {
            if (Count <= 0)
            {
                throw new Exception("集合为空");
            }
        }

        private void Check(int index, bool isAdd)
        {
            if (index < 0)
            {
                throw new Exception("索引必须为非负数");
            }
            else if (isAdd && (head <= tail && (index < head || index > tail) || head > tail && index < head && index > tail))
            {
                throw new Exception("索引越界");
            }
            else if (!isAdd && (head <= tail && (index < head || index >= tail) || head > tail && index < head && index >= tail))
            {
                throw new Exception("索引越界");
            }
        }

        private void AdjustCapacityCheck(bool isAdd)
        {
            int n = element.Length;
            if (isAdd && Count == n)
            {
                AdjustCapacity(n == 0 ? 1 : n * 2);
            }
            else if (!isAdd && Count == n / 4)
            {
                AdjustCapacity(n == 1 ? 1 : n / 2);
            }
        }

        private void AdjustCapacity(int newCapacity)
        {
            T[] old = element;
            element = new T[newCapacity];
            int mask = old.Length - 1;
            for (int i = 0; i < Count; i++)
            {
                element[i] = old[i + head & mask];
            }
            head = 0;
            tail = Count;
        }

        private struct Etor : IEnumerator<T>
        {
            private readonly ArrayDeque<T> self;
            private readonly int head;
            private readonly int tail;
            private readonly int mask;
            private int cursor;

            public Etor(ArrayDeque<T> self)
            {
                this.self = self;
                head = self.head;
                tail = self.tail;
                mask = self.element.Length - 1;
                cursor = head - 1 & mask;
            }

            public T Current
            {
                get
                {
                    if (head <= tail && (cursor < head || cursor > tail) || head > tail && cursor < head && cursor > tail)
                    {
                        throw new Exception("索引越界");
                    }
                    return self.element[cursor];
                }
            }

            public bool MoveNext()
            {
                if (cursor + 1 == tail)
                {
                    return false;
                }
                cursor = cursor + 1 & mask;
                return true;
            }
        }
    }
}